File: Scriptik/Sources/Scriptik/Services/ConfigManager.swift
import Foundation
import ServiceManagement
import SwiftUI

@Observable
@MainActor
final class ConfigManager {
    // MARK: - Config values with defaults

    var whisperModel = "medium"
    var pauseThreshold = 1.5
    var initialPrompt = ""
    var autoPaste = true
    var includeTimestamps = false
    var language = "auto"
    var whisperVenv: String
    var circlePositionX: Double = -1
    var circlePositionY: Double = -1
    var launchAtLogin: Bool = false {
        didSet {
            if launchAtLogin {
                try? SMAppService.mainApp.register()
            } else {
                try? SMAppService.mainApp.unregister()
            }
        }
    }

    // MARK: - Path constants

    static let configDir = FileManager.default.homeDirectoryForCurrentUser
        .appendingPathComponent(".config/scriptik")
    static let configFile = configDir.appendingPathComponent("config")
    static let historyDir = configDir.appendingPathComponent("history")
    static let dataDir = URL(fileURLWithPath: "/tmp/scriptik")
    static let pidFile = dataDir.appendingPathComponent("recording.pid")
    static let recordingFile = dataDir.appendingPathComponent("recording.wav")
    static let transcriptionFile = dataDir.appendingPathComponent("transcription.txt")

    // MARK: - Available options for settings UI

    static let availableModels = ["tiny", "base", "small", "medium", "large"]
    static let availableLanguages = ["auto", "en", "he"]

    // MARK: - Init

    init() {
        whisperVenv = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".local/share/scriptik/venv").path
        load()
        launchAtLogin = SMAppService.mainApp.status == .enabled
    }

    // MARK: - Load config

    func load() {
        let path = Self.configFile.path
        guard FileManager.default.fileExists(atPath: path) else { return }
        guard let contents = try? String(contentsOfFile: path, encoding: .utf8) else { return }

        for line in contents.components(separatedBy: .newlines) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            // Skip empty lines and comments
            if trimmed.isEmpty || trimmed.hasPrefix("#") { continue }

            // Split on first '='
            guard let eqIndex = trimmed.firstIndex(of: "=") else { continue }
            let key = String(trimmed[trimmed.startIndex..<eqIndex])
                .trimmingCharacters(in: .whitespaces)
            var value = String(trimmed[trimmed.index(after: eqIndex)...])
                .trimmingCharacters(in: .whitespaces)

            // Strip surrounding quotes (single or double)
            if (value.hasPrefix("\"") && value.hasSuffix("\""))
                || (value.hasPrefix("'") && value.hasSuffix("'"))
            {
                value = String(value.dropFirst().dropLast())
            }

            switch key {
            case "WHISPER_MODEL":
                whisperModel = value
            case "PAUSE_THRESHOLD":
                if let d = Double(value) { pauseThreshold = d }
            case "INITIAL_PROMPT":
                initialPrompt = value
            case "AUTO_PASTE":
                autoPaste = (value.lowercased() != "false" && value != "0")
            case "INCLUDE_TIMESTAMPS":
                includeTimestamps = (value.lowercased() != "false" && value != "0")
            case "LANGUAGE":
                language = value
            case "WHISPER_VENV":
                if !value.isEmpty { whisperVenv = value }
            case "CIRCLE_POSITION_X":
                if let d = Double(value) { circlePositionX = d }
            case "CIRCLE_POSITION_Y":
                if let d = Double(value) { circlePositionY = d }
            default:
                break
            }
        }
    }

    // MARK: - Save config

    func save() {
        let fm = FileManager.default
        let dirPath = Self.configDir.path

        if !fm.fileExists(atPath: dirPath) {
            try? fm.createDirectory(atPath: dirPath, withIntermediateDirectories: true)
        }

        var lines: [String] = []
        lines.append("# Scriptik configuration")
        lines.append("# This file is auto-generated. Manual edits are preserved.")
        lines.append("")
        lines.append("WHISPER_MODEL=\"\(whisperModel)\"")
        lines.append("PAUSE_THRESHOLD=\"\(pauseThreshold)\"")
        lines.append("INITIAL_PROMPT=\"\(initialPrompt)\"")
        lines.append("AUTO_PASTE=\"\(autoPaste)\"")
        lines.append("INCLUDE_TIMESTAMPS=\"\(includeTimestamps)\"")
        lines.append("LANGUAGE=\"\(language)\"")
        lines.append("WHISPER_VENV=\"\(whisperVenv)\"")
        lines.append("CIRCLE_POSITION_X=\"\(circlePositionX)\"")
        lines.append("CIRCLE_POSITION_Y=\"\(circlePositionY)\"")
        lines.append("")

        let content = lines.joined(separator: "\n")
        try? content.write(toFile: Self.configFile.path, atomically: true, encoding: .utf8)
    }

    // MARK: - Whisper Python path

    var whisperPython: String {
        let venvPython = (whisperVenv as NSString).appendingPathComponent("bin/python3")
        if FileManager.default.fileExists(atPath: venvPython) {
            return venvPython
        }
        // Fallback to ~/whisper-env
        let fallback = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("whisper-env/bin/python3").path
        return fallback
    }
}

File: Scriptik.Windows/Services/ConfigManager.cs
using System.ComponentModel;
using System.Globalization;
using System.IO;
using System.Runtime.CompilerServices;

namespace Scriptik.Windows.Services;

public class ConfigManager : INotifyPropertyChanged
{
    // MARK: - Config values with defaults

    private string _whisperModel = "medium";
    private double _pauseThreshold = 1.5;
    private string _initialPrompt = "";
    private bool _autoPaste = true;
    private bool _includeTimestamps;
    private string _language = "auto";
    private string _whisperVenv;
    private bool _showFloatingCircle = true;
    private bool _enableSoundFeedback = true;
    private double _circlePositionX = -1;
    private double _circlePositionY = -1;
    private bool _launchAtLogin;
    private int _hotkeyModifiers = 0x06; // MOD_CONTROL | MOD_SHIFT
    private int _hotkeyVirtualKey = 0x52; // R

    public string WhisperModel
    {
        get => _whisperModel;
        set => SetField(ref _whisperModel, value);
    }

    public double PauseThreshold
    {
        get => _pauseThreshold;
        set => SetField(ref _pauseThreshold, value);
    }

    public string InitialPrompt
    {
        get => _initialPrompt;
        set => SetField(ref _initialPrompt, value);
    }

    public bool AutoPaste
    {
        get => _autoPaste;
        set => SetField(ref _autoPaste, value);
    }

    public bool IncludeTimestamps
    {
        get => _includeTimestamps;
        set => SetField(ref _includeTimestamps, value);
    }

    public string Language
    {
        get => _language;
        set => SetField(ref _language, value);
    }

    public string WhisperVenv
    {
        get => _whisperVenv;
        set => SetField(ref _whisperVenv, value);
    }

    public bool ShowFloatingCircle
    {
        get => _showFloatingCircle;
        set => SetField(ref _showFloatingCircle, value);
    }

    public bool EnableSoundFeedback
    {
        get => _enableSoundFeedback;
        set => SetField(ref _enableSoundFeedback, value);
    }

    public double CirclePositionX
    {
        get => _circlePositionX;
        set => SetField(ref _circlePositionX, value);
    }

    public double CirclePositionY
    {
        get => _circlePositionY;
        set => SetField(ref _circlePositionY, value);
    }

    public bool LaunchAtLogin
    {
        get => _launchAtLogin;
        set
        {
            if (SetField(ref _launchAtLogin, value))
                LaunchAtLoginService.Set(value);
        }
    }

    public int HotkeyModifiers
    {
        get => _hotkeyModifiers;
        set => SetField(ref _hotkeyModifiers, value);
    }

    public int HotkeyVirtualKey
    {
        get => _hotkeyVirtualKey;
        set => SetField(ref _hotkeyVirtualKey, value);
    }

    // MARK: - Path constants

    public static string ConfigDir =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                     ".config", "scriptik");

    public static string ConfigFilePath => Path.Combine(ConfigDir, "config");
    public static string HistoryDir => Path.Combine(ConfigDir, "history");

    public static string DataDir => Path.Combine(Path.GetTempPath(), "scriptik");
    public static string PidFilePath => Path.Combine(DataDir, "recording.pid");
    public static string RecordingFilePath => Path.Combine(DataDir, "recording.wav");
    public static string TranscriptionFilePath => Path.Combine(DataDir, "transcription.txt");

    // MARK: - Available options

    public static readonly string[] AvailableModels = ["tiny", "base", "small", "medium", "large"];
    public static readonly string[] AvailableLanguages = ["auto", "en", "he"];

    // MARK: - Init

    public ConfigManager()
    {
        _whisperVenv = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "scriptik", "venv");
        Load();
        _launchAtLogin = LaunchAtLoginService.IsEnabled();
    }

    // MARK: - Whisper Python path

    public string WhisperPythonPath
    {
        get
        {
            var venvPython = Path.Combine(WhisperVenv, "Scripts", "python.exe");
            if (File.Exists(venvPython)) return venvPython;

            var fallback = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "scriptik", "venv", "Scripts", "python.exe");
            return fallback;
        }
    }

    // MARK: - Load

    public void Load()
    {
        if (!File.Exists(ConfigFilePath)) return;

        foreach (var line in File.ReadAllLines(ConfigFilePath))
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#')) continue;

            var eqIndex = trimmed.IndexOf('=');
            if (eqIndex < 0) continue;

            var key = trimmed[..eqIndex].Trim();
            var value = trimmed[(eqIndex + 1)..].Trim();

            // Strip surrounding quotes
            if (value.Length >= 2 &&
                ((value.StartsWith('"') && value.EndsWith('"')) ||
                 (value.StartsWith('\'') && value.EndsWith('\''))))
            {
                value = value[1..^1];
            }

            switch (key)
            {
                case "WHISPER_MODEL": _whisperModel = value; break;
                case "PAUSE_THRESHOLD":
                    if (double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out var pt)) _pauseThreshold = pt;
                    break;
                case "INITIAL_PROMPT": _initialPrompt = value; break;
                case "AUTO_PASTE": _autoPaste = value.ToLower() != "false" && value != "0"; break;
                case "INCLUDE_TIMESTAMPS": _includeTimestamps = value.ToLower() != "false" && value != "0"; break;
                case "LANGUAGE": _language = value; break;
                case "WHISPER_VENV":
                    if (!string.IsNullOrEmpty(value)) _whisperVenv = value;
                    break;
                case "SHOW_FLOATING_CIRCLE": _showFloatingCircle = value.ToLower() != "false" && value != "0"; break;
                case "ENABLE_SOUND_FEEDBACK": _enableSoundFeedback = value.ToLower() != "false" && value != "0"; break;
                case "CIRCLE_POSITION_X":
                    if (double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out var cx)) _circlePositionX = cx;
                    break;
                case "CIRCLE_POSITION_Y":
                    if (double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out var cy)) _circlePositionY = cy;
                    break;
                case "HOTKEY_MODIFIERS":
                    if (int.TryParse(value, out var hm)) _hotkeyModifiers = hm;
                    break;
                case "HOTKEY_VKEY":
                    if (int.TryParse(value, out var hv)) _hotkeyVirtualKey = hv;
                    break;
            }
        }
    }

    // MARK: - Save

    public void Save()
    {
        Directory.CreateDirectory(ConfigDir);

        var lines = new List<string>
        {
            "# Scriptik configuration",
            "# This file is auto-generated. Manual edits are preserved.",
            "",
            $"WHISPER_MODEL=\"{WhisperModel}\"",
            $"PAUSE_THRESHOLD=\"{PauseThreshold.ToString(CultureInfo.InvariantCulture)}\"",
            $"INITIAL_PROMPT=\"{InitialPrompt}\"",
            $"AUTO_PASTE=\"{AutoPaste}\"",
            $"INCLUDE_TIMESTAMPS=\"{IncludeTimestamps}\"",
            $"LANGUAGE=\"{Language}\"",
            $"WHISPER_VENV=\"{WhisperVenv}\"",
            $"SHOW_FLOATING_CIRCLE=\"{ShowFloatingCircle}\"",
            $"ENABLE_SOUND_FEEDBACK=\"{EnableSoundFeedback}\"",
            $"CIRCLE_POSITION_X=\"{CirclePositionX.ToString(CultureInfo.InvariantCulture)}\"",
            $"CIRCLE_POSITION_Y=\"{CirclePositionY.ToString(CultureInfo.InvariantCulture)}\"",
            $"HOTKEY_MODIFIERS=\"{HotkeyModifiers}\"",
            $"HOTKEY_VKEY=\"{HotkeyVirtualKey}\"",
            ""
        };

        File.WriteAllText(ConfigFilePath, string.Join("\n", lines));
    }

    // MARK: - INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        return true;
    }
}
